### PR TITLE
KAFKA-6486: Implemented LinkedHashMap in TimeWindows

### PR DESCRIPTION
**Pull request by asutosh936**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Thanks for the patch. We should add a test.

**Comment by guozhangwang** [comment]
> Thanks for the patch. We should add a test.

+1

**Comment by asutosh936** [comment]
@mjsax - Are the changes looks good to be merged subject to pre-merge validations?

**Comment by mjsax** [comment]
Changing the imply of a HashMap should not break anything. So yes, LGTM in general.

**Comment by ijuma** [comment]
The title seems to have a typo (`HashLinkedMap` instead of `LinkedHashMap`).

**Comment by asutosh936** [comment]
@ijuma - Updated title.

**Comment by asutosh936** [comment]
@mjsax @guozhangwang - Could someone please merge this PR if things looks good?

**Comment by mjsax** [comment]
As mentioned, the fix itself is ok, but we should add a test. Than we can merge it.

**Comment by asutosh936** [comment]
Thanks @mjsax for the updates.

**Comment by mjsax** [comment]
@asutosh936 Are you still working on adding a test for this?

**Comment by asutosh936** [comment]
@mjsax - Sorry for misunderstanding, was under impression someone from Your team is going to add the test cases. Let me know If I need to update/add test case for this fix.

**Comment by mjsax** [comment]
Oh. Sorry. I was waiting for you to add a test. It's your PR so you should add the test. Sorry for the confusion.

**Comment by asutosh936** [comment]
@mjsax - Added JUnit test case. Please review and merge.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java [hunk 1]
```diff
@@ -117,4 +123,15 @@ public void shouldNotOverlapIsOtherWindowIsAfterThisWindow() {
     public void cannotCompareTimeWindowWithDifferentWindowType() {
         window.overlap(sessionWindow);
     }
+
+    @Test
+    public void shouldComputeWindowsForLinkedHashMap() {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java]: nit: rename to `shouldReturnMatchedWindowsOrderedByTimestamp`?

**Review by guozhangwang — APPROVED** [review]
LGTM! Just left a nit comment for you. After that is addressed and jenkins has passed we can proceed and merge.

**Comment by asutosh936** [comment]
@guozhangwang - Renamed JUnit as per review.

**Comment by guozhangwang** [comment]
@asutosh936 Jenkins failures seem relevant: 

```
05:02:11 :streams:checkstyleTest[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java:22:8: Unused import - java.util.LinkedHashMap. [UnusedImports]
05:02:16 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java:24:8: Unused import - java.util.Set. [UnusedImports]
05:02:16 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java:133:46: ',' is not followed by whitespace. [WhitespaceAfter]
05:02:16 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java:134:46: ',' is not followed by whitespace. [WhitespaceAfter]
05:02:16 [ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java:135:46: ',' is not followed by whitespace. [WhitespaceAfter]
05:02:16  FAILED
```

It is because checksytle, as part of the test, reports some non-acceptable coding styles.

**Comment by asutosh936** [comment]
@guozhangwang - Updated.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java [hunk 2]
```diff
@@ -117,4 +121,15 @@ public void shouldNotOverlapIsOtherWindowIsAfterThisWindow() {
     public void cannotCompareTimeWindowWithDifferentWindowType() {
         window.overlap(sessionWindow);
     }
+
+    @Test
+    public void shouldReturnMatchedWindowsOrderedByTimestamp() {
+        TimeWindows windows = TimeWindows.of(12L).advanceBy(5L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java]: Nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java [hunk 3]
```diff
@@ -117,4 +121,15 @@ public void shouldNotOverlapIsOtherWindowIsAfterThisWindow() {
...
+        Map<Long, TimeWindow> matched = windows.windowsFor(21L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java]: Nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java [hunk 4]
```diff
@@ -117,4 +121,15 @@ public void shouldNotOverlapIsOtherWindowIsAfterThisWindow() {
...
+
+        Long[] expected = matched.keySet().toArray(new Long[matched.size()]);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/TimeWindowTest.java]: Nit: add `final`

**Review by mjsax — COMMENTED** [review]
Thanks to adding the unit test. Just some code style nits (we use `final` whenever possible).

**Comment by asutosh936** [comment]
@mjsax - Implemented review comments

**Review by mjsax — APPROVED** [review]
Thanks for the PR @asutosh936 !

LGTM. Merging to `trunk`

**Pull request merged**